### PR TITLE
Obiektowość #4 Cars assgiment

### PR DESCRIPTION
**Pull request by marcindus**
Fixed interfaces for :   constructors: Petrol,Electric and Hybrid and Engine.accelerate

### CODE REVIEW

File: PetrolEngine.cpp [hunk 1]
```diff
@@ -1,20 +1,34 @@
...
+void PetrolEngine::changeGear(int gear) {
+  // TODO: Add checking if gear is between -1 and gears_
+  // -1 is for REAR
+  // 0 is for NEUTRAL
+
+  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+            << std::endl;
+  if (abs(gear - currentGear_) != 1 or gear < -1 or gear > gears_.gears_) {
+    throw InvalidGear("Zgrzyyyyt!");
+  }
```
Comment (WBigaj) [inline PetrolEngine.cpp:21]: I suggest making '-1' and '1' constexprs instead of magic values.

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

File: PetrolEngine.cpp (hunk 1 above)
Comment (MaciejSalamonski) [reply PetrolEngine.cpp:21]: +1

**Review by MaciejSalamonski — APPROVED** [review]
Code looks good ;p

**Review by jakjan95 — COMMENTED** [review]
Code looks good but I am not sure that creating IEngine is the best option for this project. This can open up the possibility of an engine misassignment and additional logic will be needed to implement.

File: ElectricCar.cpp [hunk 2]
```diff
@@ -1,16 +1,48 @@
...
+            << std::endl;
+}
+void ElectricCar::brake() const {
+  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+            << std::endl;
+}
+void ElectricCar::accelerate(Speed speed) const {
+  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+            << std::endl;
+}
```
Comment (ziobron) [inline ElectricCar.cpp:29]: you may move this behavior into ICar and use the common implementation. Code duplication is something that we want to avoid.

**Review by ziobron — CHANGES_REQUESTED** [review]
There are some code duplications that can be avoided. 
You have tested only PetrolEngine and PetrolCar classes. Tests for Electric and Hybrid car (or parametrized tests for each car) would be welcomed :)

File: ICar.hpp [hunk 3]
```diff
@@ -1,17 +1,50 @@
...
+
+  virtual void refuel() {
+    std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+              << std::endl;
+  }
+
+  virtual void charge() {
+    std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+              << std::endl;
+  }
```
Comment (ziobron) [inline ICar.hpp:49]: these 2 functions should not be virtual nor should they be in an interface. ICar interface has `fill()` method that should be implemented in ElectricCar to call `charge()` and in PetrolCar to call `refuel()`.

File: ElectricCar.cpp [hunk 4]
```diff
@@ -2,47 +2,15 @@
...
-            << std::endl;
-  engine_->stop();
-}
+ElectricCar::~ElectricCar() {}
+void ElectricCar::turnLeft() const {}
+void ElectricCar::turnRight() const {}
+void ElectricCar::brake() const {}
+void ElectricCar::accelerate(Speed speed) const {}
+void ElectricCar::start_engine() const { engine_->start(); }
+void ElectricCar::stop_engine() const { engine_->stop(); }
```
Comment (ziobron) [inline ElectricCar.cpp:13]: that's also something we do not want. You are overriding those methods with empty implementations. You should use your base class implementation, so do not override them. Simply remove them from cpp and hpp. The same for other types of cars.

File: test.cpp [hunk 5]
```diff
@@ -20,12 +20,30 @@ class CarTestSuite : public ::testing::Test {
   PetrolCar m_sut;
 };
 
-TEST_F(CarTestSuite, stopEngineShouldCallStopOEngine) {
+
+class HybridCarTestSuite : public ::testing::Test {
```
Comment (ziobron) [inline test.cpp:24]: It's a good practice to move tests for hybrid car into another file

File: test.cpp [hunk 6]
```diff
@@ -44,5 +62,17 @@ TEST(EngineTestSuite, ProperGear) {
...
+  EXPECT_CALL(*m_engineMock_electric_rawPtr, stop());
+  m_sut.stop_engine();
+}
+
+TEST_F(HybridCarTestSuite, startEngineShouldCallStopOEngine) {
+  EXPECT_CALL(*m_engineMock_petrol_rawPtr, start());
+  EXPECT_CALL(*m_engineMock_electric_rawPtr, start());
+  m_sut.start_engine();
+}
+
```
Comment (ziobron) [inline test.cpp:78]: Electric car tests are still missing.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: test.cpp (hunk 6 above)
Comment (marcindus) [reply test.cpp:78]: Added tests. Still some tests missing But is late....  I will add them I promise,.. maybe even parametrized.

Comment (ziobron) [reply test.cpp]: For me it's good enough :)
I know that you can write other similar tests.

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
3 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522] }

**Pull request closed without merging**